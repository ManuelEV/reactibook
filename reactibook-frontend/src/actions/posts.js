import Swal from 'sweetalert2';
import { fileUpload } from '../helpers/fileUpload';
import { loadPosts } from "../helpers/loadPosts";
import { postOptions, putOptions } from '../helpers/requestOptions';
import { types } from "../types/types";


export const startNewPost = (content, visibility) => {
    return async( dispatch, getState ) => {
        const { uid, name } = getState().auth;

        const { currentFileUrl } = getState().posts;

        const newPost = {
            author: {id: uid, name},
            content,
            visibility,
            date: new Date().getTime(),
            fileUrl: currentFileUrl,
        };
        
        const options = postOptions(newPost);

        const response = await fetch(`${process.env.REACT_APP_API_URL}/posts`, options);

        const body = await response.json();

        const newPostId = body.post.id;

        dispatch(addNewPost({id: newPostId, ...newPost}));

    }
}

export const addNewPost = (newPost) => ({
    type: types.postsAddNew,
    payload: {
        ...newPost
    }
})

export const startLoadingPosts = () => {
    return async (dispatch) => {
        const posts = await loadPosts();
        console.log('Loading posts... ', posts)
        dispatch(setPosts(posts));
    }
}

export const setPosts = (posts) => ({

    type: types.postsLoad,
    payload: posts,

})

export const startUpdatingPost = (id, data) => {
    return async(dispatch, getState) => {

        const options = putOptions(data);

        const response = await fetch(`${process.env.REACT_APP_API_URL}/posts/${id}`, options);

        const body = await response.json();

        console.log(body);

        const editedPost = body.post;

       dispatch(refreshPost(editedPost));
        Swal.fire('Saved', 'El post se guardó con éxito', 'success');
    }
}

export const refreshPost = (post) => ({
    type: types.postsUpdated,
    payload: post
});


export const updateCurrentFileUrl = (fileUrl) => ({
    type: types.postsCurrentFileUrl,
    payload: fileUrl
})


export const startUploading = (file) => {
    return async (dispatch, getState) => {

        Swal.fire({
            title: 'Uploading',
            text: 'Please wait',
            allowOutsideClick: false,
            showConfirmButton: false,
        });
        Swal.showLoading();

        const fileUrl = await fileUpload(file);

        dispatch(updateCurrentFileUrl(fileUrl));
        
        Swal.close();

        console.log(fileUrl);

    }
}




export const startDeleting = (id) => {
    return async (dispatch, getState) => {

        const options = {
            method: 'DELETE',
        };

        const response = await fetch(`${process.env.REACT_APP_API_URL}/posts/${id}`, options);

        const body = await response.json();

        console.log(body);

        dispatch(deletePost(id));

    }
}

export const deletePost = (id) => ({
    type: types.postsDelete,
    payload: id,
})


export const postLogout = () => ({
    type: types.postsLogoutCleaning,
})