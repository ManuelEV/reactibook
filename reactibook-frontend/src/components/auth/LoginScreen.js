import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import validator from 'validator';
import { Link } from 'react-router-dom'
import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth';
import { removeLoginError, setLoginError } from '../../actions/ui';
import { useForm } from '../../hooks/useForm';

export const LoginScreen = () => {

    const dispatch = useDispatch();

    const { loading } = useSelector( state => state.ui );

    const [formValues, handleInputChange] = useForm({
        email: '',
        password: '',
    });

    const {email, password} = formValues;

    const handleLogin = (e) => {
        e.preventDefault();
        if(isFormValid()){
            dispatch(startLoginEmailPassword(email, password));
        }
    };

    const handleGoogleLogin = () => {
        dispatch(startGoogleLogin());
    }

    
    const isFormValid = () => {
        const errors = {};
        let isValid = true;


        if ( !validator.isEmail( email ) ){
            errors['emailMsg'] = 'El email no es válido';
            isValid = false;
        }
        
        if (password.length < 5 ){
            errors['passwordMsg'] = 'La contraseña debe ser de al menos 6 caracteres';
            isValid = false;
        }

        if (isValid) {
            dispatch( removeLoginError() );
        } else {
            dispatch(setLoginError(errors));
        }

        return isValid;
    };

    const { emailMsg, passwordMsg } = useSelector( state => state.ui.loginErrors );

    return (
        <div className="w-screen h-screen flex justify-center items-center text-white bg-gray-800">

            <form 
                className="md:w-1/4 sm:w-1/2 bg-gradient-to-b from-primary-400 to-primary-800
                rounded-lg shadow-lg p-10"
                onSubmit={handleLogin}
            >

                <div className="">
                    Email:
                </div>

                <div className="w-full">
                    <input 
                        type="text"
                        placeholder="Email"
                        name="email"
                        className="w-full rounded px-2 text-black
                        border border-transparent focus:outline-none focus:ring-2 
                        focus:ring-secondary-100 focus:border-transparent"
                        autoComplete="off"
                        value={email}
                        onChange={handleInputChange}
                    />
                    <span className="text-sm text-red-500">
                        {emailMsg}
                    </span>
                </div>

                <div className="pt-2">
                    Password:
                </div>

                <div className="w-full">
                    <input 
                        type="password"
                        placeholder="Password"
                        name="password"
                        className="w-full rounded px-2 text-black
                        border border-transparent focus:outline-none focus:ring-2 
                        focus:ring-secondary-900 focus:border-transparent"
                        value={password}
                        onChange={handleInputChange}
                    />
                    <span className="text-sm text-red-500">
                        {passwordMsg}
                    </span>
                </div>


                <div className="w-full flex justify-center my-3">
                    <button
                        type="submit"
                        className="bg-secondary-700 py-1 px-5 rounded shadow font-bold
                            hover:bg-secondary-900"
                        disabled={loading}
                    >
                        Login
                    </button>
                </div>

                
                <div className="w-full mt-4">

                    <div 
                        className="w-full cursor-pointer"
                        onClick={handleGoogleLogin}
                    >
                        <div className="flex justify-center">
                            <img 
                                className="h-14 w-14 rounded-full p-1 bg-black bg-opacity-20" 
                                src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" 
                                alt="google button" 
                            />
                        </div>
                        <div className="text-center">
                            <span>Ingresar con Google</span>
                        </div>
                    </div>
                </div>

                <div className="w-full text-center cursor-pointer underline mt-4">
                    <Link 
                        to="/auth/register"
                        className=""
                    >
                        Crear una nueva cuenta  
                    </Link>
                </div>

            </form>
        </div>
    )
}
