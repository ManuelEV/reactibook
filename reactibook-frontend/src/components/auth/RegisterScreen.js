import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import validator from 'validator';
import { removeError, setError } from '../../actions/ui';
import { startRegisterWithEmailPasswordName } from '../../actions/auth';
import { useForm } from '../../hooks/useForm';

export const RegisterScreen = () => {

    const dispatch = useDispatch();

    const { msgError } = useSelector(state => state.ui);


    const [formValues, handleInputChange] = useForm({
        name: '', 
        email: '', 
        password: '', 
        passwordConfirmation: '',
    });

    const {name, email, password, passwordConfirmation} = formValues;

    const handleRegister = (e) => {
        e.preventDefault();

        if (isFormValid()){
            dispatch(startRegisterWithEmailPasswordName(email, password, name));
        }

    };

    const isFormValid = () => {
        if (name.trim().length === 0){
            dispatch(setError('El nombre es requerido'));
            return false;
        } else if ( !validator.isEmail( email ) ){
            dispatch(setError('El email no es válido'));
            return false;
        }else if ( password !== passwordConfirmation || password.length < 5 ){
            dispatch(setError('La contraseña debe ser de al menos 6 caracteres'));
            return false;
        }

        dispatch( removeError() );

        return true;
    };


    return (
        <div className="w-screen h-screen flex justify-center items-center bg-gray-800 text-white">

            
            <form 
                className="md:w-1/4 sm:w-1/2 bg-gradient-to-b from-primary-400 to-primary-800
                rounded-lg shadow-lg p-10"
                onSubmit={handleRegister}
            >

                {
                    msgError &&
                    (
                        <div className="w-full text-sm text-red-500 text-center">
                            {msgError}
                        </div>
                    )
                }

                <div className="w-full my-2">
                    <input 
                        type="text"
                        placeholder="Name"
                        name="name"
                        className="w-full rounded px-2 text-black
                        border border-transparent focus:outline-none focus:ring-2 
                        focus:ring-secondary-100 focus:border-transparent"
                        autoComplete="off"
                        value={name}
                        onChange={handleInputChange}
                    />
                </div>

                <div className="w-full my-2">
                    <input 
                        type="text"
                        placeholder="Email"
                        name="email"
                        className="w-full rounded px-2 text-black
                        border border-transparent focus:outline-none focus:ring-2 
                        focus:ring-secondary-100 focus:border-transparent"
                        autoComplete="off"
                        value={email}
                        onChange={handleInputChange}
                    />
                </div>

                <div className="w-full my-2">
                    <input 
                        type="password"
                        placeholder="Password"
                        name="password"
                        className="w-full rounded px-2 text-black
                        border border-transparent focus:outline-none focus:ring-2 
                        focus:ring-secondary-100 focus:border-transparent"
                        value={password}
                        onChange={handleInputChange}
                    />
                </div>

                <div className="w-full my-2">
                    <input 
                        type="password"
                        placeholder="Confirm password"
                        name="passwordConfirmation"
                        className="w-full rounded px-2 text-black
                        border border-transparent focus:outline-none focus:ring-2 
                        focus:ring-secondary-100 focus:border-transparent"
                        value={passwordConfirmation}
                        onChange={handleInputChange}
                    />
                </div>


                <div className="w-full flex justify-center my-4">
                    <button
                        type="submit"
                        className="bg-secondary-700 py-1 px-5 rounded shadow font-bold
                        hover:bg-secondary-900"
                    >
                        Registrarse
                    </button>
                </div>



                <div className="w-full text-center cursor-pointer">
                    <Link 
                        to="/auth/login"
                        className="link"
                    >
                        ¿Ya estás registrado?
                    </Link>
                </div>

            </form>
            
        </div>
    )
}
