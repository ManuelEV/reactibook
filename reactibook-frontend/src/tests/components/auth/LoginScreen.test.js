import React from 'react';
import { mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';

import { LoginScreen } from '../../../components/auth/LoginScreen';
import { startGoogleLogin, startLoginEmailPassword } from '../../../actions/auth';

jest.mock('../../../actions/auth', () => ({
    startGoogleLogin: jest.fn(),
    startLoginEmailPassword: jest.fn()
}))

const middlewares = [thunk];

const mockStore = configureStore(middlewares);

const initState = {
    auth: {},
    ui: {
        loading: false,
        msgError: null,
        loginErrors: {
            emailMsg: '',
            passwordMsg: '',
        },
    },
};

let store = mockStore(initState);

store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={store}>
        <MemoryRouter>
            <LoginScreen />
        </MemoryRouter>
    </Provider>
);

describe('Tests in <LoginScreen />', () => {

    beforeEach(() => {
        store = mockStore(initState);
        jest.clearAllMocks();
    });
    
    
    test('snapshot should match', () => {
        expect(wrapper).toMatchSnapshot();
    });
    
    test('should run startGoogleLogin action', () => {
        
        // find by class and simulate click

        wrapper.find('.w-full .cursor-pointer').prop('onClick')();

        expect(startGoogleLogin).toHaveBeenCalled();

    });

    test('It shouldnt run the startLoginEmailPassword action if the inputs dont exist', () => {
        
        // find form and simulate onSubmit

        wrapper.find('form').prop('onSubmit')({preventDefault(){}});

        // we expect isValid is false (no params)

        expect(startLoginEmailPassword).not.toHaveBeenCalled();

    });
    

})

