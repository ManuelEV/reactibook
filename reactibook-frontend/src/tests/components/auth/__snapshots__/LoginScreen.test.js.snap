// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <LoginScreen /> snapshot should match 1`] = `
<div
  className="w-screen h-screen flex justify-center items-center text-white bg-gray-800"
>
  <form
    className="md:w-1/4 sm:w-1/2 bg-gradient-to-b from-primary-400 to-primary-800 rounded-lg shadow-lg p-10"
    onSubmit={[Function]}
  >
    <div
      className=""
    >
      Email:
    </div>
    <div
      className="w-full"
    >
      <input
        autoComplete="off"
        className="w-full rounded px-2 text-black border border-transparent focus:outline-none focus:ring-2  focus:ring-secondary-100 focus:border-transparent"
        name="email"
        onChange={[Function]}
        placeholder="Email"
        type="text"
        value=""
      />
      <span
        className="text-sm text-red-500"
      />
    </div>
    <div
      className="pt-2"
    >
      Password:
    </div>
    <div
      className="w-full"
    >
      <input
        className="w-full rounded px-2 text-black border border-transparent focus:outline-none focus:ring-2  focus:ring-secondary-900 focus:border-transparent"
        name="password"
        onChange={[Function]}
        placeholder="Password"
        type="password"
        value=""
      />
      <span
        className="text-sm text-red-500"
      />
    </div>
    <div
      className="w-full flex justify-center my-3"
    >
      <button
        className="bg-secondary-700 py-1 px-5 rounded shadow font-bold hover:bg-secondary-900"
        disabled={false}
        type="submit"
      >
        Login
      </button>
    </div>
    <div
      className="w-full mt-4"
    >
      <div
        className="w-full cursor-pointer"
        onClick={[Function]}
      >
        <div
          className="flex justify-center"
        >
          <img
            alt="google button"
            className="h-14 w-14 rounded-full p-1 bg-black bg-opacity-20"
            src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
          />
        </div>
        <div
          className="text-center"
        >
          <span>
            Ingresar con Google
          </span>
        </div>
      </div>
    </div>
    <div
      className="w-full text-center cursor-pointer underline mt-4"
    >
      <a
        className=""
        href="/auth/register"
        onClick={[Function]}
      >
        Crear una nueva cuenta
      </a>
    </div>
  </form>
</div>
`;
